src/components/GifTile.js

```diff
@@ -0,0 +1,127 @@
+import React, { Component } from 'react'
+import { NavLink } from 'react-router-dom'
+import styled from 'styled-components';
+import { connect } from 'react-redux'
+
+class GifTile extends Component {
+  constructor(props) {
+    super(props);
+    this.showControls = this.showControls.bind(this);
+    this.hideControls = this.hideControls.bind(this);
+    this.save = this.save.bind(this);
+    this.remove = this.remove.bind(this);
+    this.state = {
+      showControls: false
+    };
+  }
+
+  render () {
+    return (
+      <Tile onMouseOver={this.showControls} onMouseLeave={this.hideControls}>
+        {
+          this.props.isFavorite(this.props.data.id) &&
+          <IconFavorite>
+            <span role="img" aria-label="favorite">⭐</span>
+          </IconFavorite>
+        }
+        <NavLink to={'/details/' + this.props.data.id}>
+          <ImageStyled src={this.props.data.images.fixed_width.url} alt={this.props.data.slug}/>
+        </NavLink>
+        {
+          this.state.showControls && !this.props.isFavorite(this.props.data.id) &&
+          <Controls>
+            <ButtonStyled onClick={this.save}>Save to favorites</ButtonStyled>
+          </Controls>
+        }
+        {
+          this.state.showControls && this.props.isFavorite(this.props.data.id) &&
+          <Controls>
+            <ButtonStyledRed onClick={this.remove}>Remove from favorites</ButtonStyledRed>
+          </Controls>
+        }
+      </Tile>
+    )
+  }
+
+  showControls () {
+    this.setState({
+      showControls: true
+    });
+  }
+
+  hideControls () {
+    this.setState({
+      showControls: false
+    });
+  }
+
+  save () {
+    this.props.saveFavorite(this.props.data)
+  }
+
+  remove () {
+    this.props.removeFavorite(this.props.data.id)
+  }
+}
+
+function mapStateToProps(state) {
+  return {
+    isFavorite: (id) => {
+      return !!state.favoriteGifs.find(gif => gif.id === id)
+    }
+  };
+}
+
+function mapDispatchToProps (dispatch) {
+  return {
+    saveFavorite: (gifData) => dispatch({ type: 'SAVE_FAVORITE', gifData }),
+    removeFavorite: (gifId) => dispatch({ type: 'REMOVE_FAVORITE', gifId })
```

Comment:
You can use `bindActionCreators` to avoid the manual definition of these functions.